### PR TITLE
Less props passed to ABProvider

### PR DESCRIPTION
**Pull request by mxdvl**
## What does this change?

Add a method to filter out unused switches.

## Why?

We should only pass required props to `Island` children.

### Before

> `{"sectionName":"uk-news","palette":{"text":{"headline":"#121212","seriesTitle":"#C70000","sectionTitle":"#C70000","byline":"#C70000","twitterHandle":"#707070","caption":"#707070","captionLink":"#C70000","subMeta":"#C70000","subMetaLabel":"#707070","subMetaLink":"#707070","syndicationButton":"#707070","articleLink":"#C70000","articleLinkHover":"#C70000","cardHeadline":"#121212","cardByline":"#C70000","cardKicker":"#C70000","linkKicker":"#C70000","cardStandfirst":"#121212","cardFooter":"#707070","headlineByline":"#C70000","standfirst":"#121212","standfirstLink":"#C70000","branding":"#C70000","disclaimerLink":"#AB0613","signInLink":"#AB0613","richLink":"#C70000","pullQuote":"#AB0613","pullQuoteAttribution":"#C70000","witnessIcon":"#C70000","witnessAuthor":"#C70000","witnessTitle":"#C70000","carouselTitle":"#C70000","calloutHeading":"#007ABC","dropCap":"#AB0613","blockquote":"#707070","numberedTitle":"#C70000","numberedPosition":"#707070","overlayedCaption":"#FFFFFF","shareCount":"#707070","shareCountUntilDesktop":"#707070"},"background":{"article":"transparent","seriesTitle":"transparent","sectionTitle":"transparent","avatar":"#FF5943","card":"#F6F6F6","headline":"transparent","headlineByline":"transparent","bullet":"#C70000","bulletStandfirst":"#DCDCDC","header":"transparent","standfirst":"transparent","richLink":"#C70000","imageTitle":"#C70000","speechBubble":"#C70000","carouselDot":"#C70000","carouselDotFocus":"#C70000","headlineTag":"#AB0613","mostViewedTab":"#AB0613","matchNav":"#FFE500","analysisUnderline":"rgba(199, 0, 0, 0.5)"},"fill":{"commentCount":"#C70000","commentCountUntilDesktop":"#C70000","shareIcon":"#C70000","shareCountIcon":"#707070","shareCountIconUntilDesktop":"#707070","shareIconGrayBackground":"#AB0613","cameraCaptionIcon":"#707070","richLink":"#C70000","quoteIcon":"#C70000","blockquoteIcon":"#C70000"},"border":{"syndicationButton":"#DCDCDC","subNav":"#C70000","articleLink":"#DCDCDC","articleLinkHover":"#C70000","liveBlock":"#C70000","standfirstLink":"#DCDCDC","headline":"#DCDCDC","standfirst":"#DCDCDC","richLink":"#C70000","navPillar":"#FF5943","article":"#DCDCDC","lines":"#DCDCDC","matchTab":"#DCDCDC","activeMatchTab":"#005689"},"topBar":{"card":"#C70000"},"hover":{"headlineByline":"#AB0613","standfirstLink":"#DCDCDC"}},"ajaxUrl":"https://api.nextgen.guardianapps.co.uk","switches":{"anniversaryHeaderSvg":true,"abSpacefinderOkr3RichLinks":true,"prebidAppnexusUkRow":true,"commercialMetrics":true,"prebidTrustx":true,"scAdFreeBanner":false,"abSpacefinderOkr2ImagesLoaded":false,"prebidPermutiveAudience":true,"compareVariantDecision":false,"enableSentryReporting":true,"lazyLoadContainers":true,"adFreeStrictExpiryEnforcement":false,"liveblogRendering":true,"remarketing":true,"fetchNonRefreshableLineItems":true,"registerWithPhone":false,"targeting":true,"remoteHeader":true,"extendedMostPopularFronts":true,"slotBodyEnd":true,"prebidImproveDigitalSkins":true,"emailInlineInFooter":true,"showNewPrivacyWordingOnEmailSignupEmbeds":true,"facebookTrackingPixel":true,"iasAdTargeting":true,"extendedMostPopular":true,"prebidAnalytics":true,"prebidCriteo":true,"puzzlesBanner":true,"imrWorldwide":true,"acast":true,"twitterUwt":true,"prebidAppnexusInvcode":true,"a9HeaderBidding":true,"prebidAppnexus":true,"enableDiscussionSwitch":true,"standaloneCommercialBundle":true,"prebidXaxis":true,"abSpacefinderOkr1FilterNearby":false,"stickyVideos":true,"interactiveFullHeaderSwitch":true,"discussionAllPageSize":true,"prebidUserSync":true,"audioOnwardJourneySwitch":true,"mobileStickyPrebid":true,"externalVideoEmbeds":true,"simpleReach":true,"carrotTrafficDriver":true,"sentinelLogger":true,"geoMostPopular":true,"weAreHiring":true,"relatedContent":true,"thirdPartyEmbedTracking":true,"prebidOzone":true,"ampAmazon":true,"prebidAdYouLike":true,"mostViewedFronts":true,"abSignInGateMainControl":true,"ampPrebid":true,"googleSearch":true,"brazeSwitch":true,"consentManagement":true,"commercial":true,"redplanetForAus":true,"prebidSonobi":true,"idProfileNavigation":true,"confiantAdVerification":true,"discussionAllowAnonymousRecommendsSwitch":false,"scrollDepth":true,"permutive":true,"comscore":true,"webFonts":true,"prebidImproveDigital":true,"ophan":true,"crosswordSvgThumbnails":true,"prebidTriplelift":true,"weather":true,"commercialOutbrainNewids":true,"abSignInGateMainVariant":true,"abAdblockAsk":true,"prebidPubmatic":true,"serverShareCounts":false,"autoRefresh":true,"enhanceTweets":true,"prebidIndexExchange":true,"prebidOpenx":true,"prebidHeaderBidding":true,"idCookieRefresh":true,"sharingComments":true,"discussionPageSize":true,"smartAppBanner":false,"boostGaUserTimingFidelity":false,"historyTags":true,"mobileStickyLeaderboard":true,"surveys":true,"remoteBanner":true,"emailSignupRecaptcha":true,"prebidSmart":true,"inizio":true},"pageIsSensitive":false,"isDev":false}`

### After

> `{"sectionName":"uk-news","palette":{"text":{"headline":"#121212","seriesTitle":"#C70000","sectionTitle":"#C70000","byline":"#C70000","twitterHandle":"#707070","caption":"#707070","captionLink":"#C70000","subMeta":"#C70000","subMetaLabel":"#707070","subMetaLink":"#707070","syndicationButton":"#707070","articleLink":"#C70000","articleLinkHover":"#C70000","cardHeadline":"#121212","cardByline":"#C70000","cardKicker":"#C70000","linkKicker":"#C70000","cardStandfirst":"#121212","cardFooter":"#707070","headlineByline":"#C70000","standfirst":"#121212","standfirstLink":"#C70000","branding":"#C70000","disclaimerLink":"#AB0613","signInLink":"#AB0613","richLink":"#C70000","pullQuote":"#AB0613","pullQuoteAttribution":"#C70000","witnessIcon":"#C70000","witnessAuthor":"#C70000","witnessTitle":"#C70000","carouselTitle":"#C70000","calloutHeading":"#007ABC","dropCap":"#AB0613","blockquote":"#707070","numberedTitle":"#C70000","numberedPosition":"#707070","overlayedCaption":"#FFFFFF","shareCount":"#707070","shareCountUntilDesktop":"#707070"},"background":{"article":"transparent","seriesTitle":"transparent","sectionTitle":"transparent","avatar":"#FF5943","card":"#F6F6F6","headline":"transparent","headlineByline":"transparent","bullet":"#C70000","bulletStandfirst":"#DCDCDC","header":"transparent","standfirst":"transparent","richLink":"#C70000","imageTitle":"#C70000","speechBubble":"#C70000","carouselDot":"#C70000","carouselDotFocus":"#C70000","headlineTag":"#AB0613","mostViewedTab":"#AB0613","matchNav":"#FFE500","analysisUnderline":"rgba(199, 0, 0, 0.5)"},"fill":{"commentCount":"#C70000","commentCountUntilDesktop":"#C70000","shareIcon":"#C70000","shareCountIcon":"#707070","shareCountIconUntilDesktop":"#707070","shareIconGrayBackground":"#AB0613","cameraCaptionIcon":"#707070","richLink":"#C70000","quoteIcon":"#C70000","blockquoteIcon":"#C70000"},"border":{"syndicationButton":"#DCDCDC","subNav":"#C70000","articleLink":"#DCDCDC","articleLinkHover":"#C70000","liveBlock":"#C70000","standfirstLink":"#DCDCDC","headline":"#DCDCDC","standfirst":"#DCDCDC","richLink":"#C70000","navPillar":"#FF5943","article":"#DCDCDC","lines":"#DCDCDC","matchTab":"#DCDCDC","activeMatchTab":"#005689"},"topBar":{"card":"#C70000"},"hover":{"headlineByline":"#AB0613","standfirstLink":"#DCDCDC"}},"ajaxUrl":"https://api.nextgen.guardianapps.co.uk","abTestSwitches":{"abSpacefinderOkr3RichLinks":true,"abSpacefinderOkrMegaTest":false,"abSignInGateMainControl":true,"abSignInGateMainVariant":true,"abAdblockAsk":true},"pageIsSensitive":false,"isDev":false}`
```

### CODE REVIEW

File: dotcom-rendering/src/web/lib/filterSwitches.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+export type ABTestSwitches = Record<`ab${string}`, boolean>;
+
+export const filterABTestSwitches = (switches: Switches): ABTestSwitches =>
+	Object.fromEntries(
+		Object.entries(switches).filter(([key]) => key.startsWith('ab')),
```
Comment (oliverlloyd) [inline dotcom-rendering/src/web/lib/filterSwitches.ts:5]: Is there a convention that switches used for ab tests are given the `ab` prefix and this function is built on top of that pattern?

**Review by oliverlloyd — COMMENTED** [review]
Are we saying here that within `switches` there is actually a subset of switches used for ab tests and that sometimes in the code these are getting confused with the main switches used as feature flags? Does this perhaps surface a larger problem that we should really have two properties on the model?

If the above is true, then ideally this transformation would happen earlier in the stack, in which case this change could be made into [an enhancer](https://github.com/guardian/dotcom-rendering/blob/150fc2d81e6a66d9c3336185e874fc8cd0288546/dotcom-rendering/docs/patterns/enhance-capi.md). We use enhancers for the types of model changes that we:

1) want to happen now so that we can write the rendering code the way we think it should be be but
2) also want to move up to a higher level in the stack

By using enhancers we make any transition up in future easier (because the mutation is contained) and we also make writing rendering code easier as the model fits the right pattern.

File: dotcom-rendering/src/web/lib/filterSwitches.ts (hunk 1 above)
Comment (mxdvl) [reply dotcom-rendering/src/web/lib/filterSwitches.ts:5]: Yes, this is [a convention followed in frontend](https://github.com/guardian/frontend/blob/016e9e268b47f2fef04f491af12d7e2c9577a2fe/common/app/conf/switches/ABTestSwitches.scala) and [the ab-testing framework expects it](https://github.com/guardian/ab-testing/blob/b5de0e09259d99399f192e4696470812b2aa45b1/packages/ab-core/src/core.ts#L29-L30).

Sadly I couldn’t find documentation outside of the actual LoC where this stuff is evaluated. I tried to bring a change to the ab-testing library to use a template string to indicate this fact with `ab${string}`, but couldn’t get the repo to behave.

**Comment by OllysCoding** [comment]
> in which case this change could be made into [an enhancer](https://github.com/guardian/dotcom-rendering/blob/150fc2d81e6a66d9c3336185e874fc8cd0288546/dotcom-rendering/docs/patterns/enhance-capi.md)

Worth keeping in mind that we recently refactored `enhanceCAPI` into `enhanceBlocks` to support the `/Blocks` endpoint on DCR.

If we wanted to use an enhance pattern for this solution, we'd need to re-introduce some kind of enhancement protocol which works for the wider CAPI object, while making sure it doesn't prevent support for our new endpoints like `/Blocks` and `/KeyEvents`.

**Comment by oliverlloyd** [comment]
> 

We already have [enhanceStandfirst](https://github.com/guardian/dotcom-rendering/blob/3597fce007af9f20c0c0f1b7d38c33a060613d89/dotcom-rendering/src/model/enhanceStandfirst.ts) which I think follows the pattern we would need here. 

Right now we have

```
		const CAPI = {
			...data,
			blocks: enhanceBlocks(data.blocks, data.format),
			standfirst: enhanceStandfirst(data.standfirst),
		};
```

So maybe we could

```
		const CAPI = {
			...data,
			blocks: enhanceBlocks(data.blocks, data.format),
			standfirst: enhanceStandfirst(data.standfirst),
			abThings: addAbThings(data.switches),
			switches: enhanceSwitches(data.switches),
		};
```

or something like that?

**Comment by mxdvl** [comment]
In light of #4200, I think #4403 is a better approach.

**Pull request closed without merging**